xtask/src/flash_image.rs

```diff
@@ -0,0 +1,504 @@
+        buffer.extend_from_slice(&self.image_count.to_le_bytes());
+        buffer
+    }
+}
+
+impl FlashImagePayload {
+    pub fn serialize(&self) -> Vec<u8> {
+        let mut buffer = Vec::new();
+        for info in &self.image_info {
+            buffer.extend_from_slice(&info.identifier.to_le_bytes());
+            buffer.extend_from_slice(&info.image_offset.to_le_bytes());
+            buffer.extend_from_slice(&info.size.to_le_bytes());
+        }
+        for image in &self.images {
+            buffer.extend_from_slice(&image.data);
+        }
+        buffer
+    }
+}
+
+impl FlashImageChecksum {
+    pub fn new(header: &FlashImageHeader, payload: &FlashImagePayload) -> Self {
+        Self {
+            header: Self::calculate_checksum(&header.serialize()),
+            payload: Self::calculate_checksum(&payload.serialize()),
+        }
+    }
+    pub fn calculate_checksum(data: &[u8]) -> u32 {
+        let mut hasher = Hasher::new();
+        hasher.update(data);
+        hasher.finalize()
+    }
+}
+
+pub(crate) fn flash_image_create(
+    caliptra_fw_path: &str,
+    soc_manifest_path: &str,
+    mcu_runtime_path: &str,
+    soc_image_paths: &Option<Vec<String>>,
+    output_path: &str,
+) -> Result<(), DynError> {
+    let mut images: Vec<FirmwareImage> = Vec::new();
+    images.push(FirmwareImage::new(
+        CALIPTRA_FMC_RT_IDENTIFIER,
+        caliptra_fw_path,
+    )?);
+    images.push(FirmwareImage::new(
+        SOC_MANIFEST_IDENTIFIER,
+        soc_manifest_path,
+    )?);
+    images.push(FirmwareImage::new(MCU_RT_IDENTIFIER, mcu_runtime_path)?);
+    if let Some(soc_image_paths) = soc_image_paths {
+        let mut soc_image_identifer = SOC_IMAGES_BASE_IDENTIFIER;
+        for soc_image_path in soc_image_paths {
+            images.push(FirmwareImage::new(soc_image_identifer, soc_image_path)?);
+            soc_image_identifer += 1;
+        }
+    }
+
+    let flash_image = FlashImage::new(&mut images);
+    flash_image.write_to_file(output_path)?;
+
+    Ok(())
+}
+
+pub(crate) fn flash_image_verify(image_file_path: &str) -> Result<(), DynError> {
+    let mut file = File::open(image_file_path).map_err(|e| {
+        Error::new(
+            ErrorKind::NotFound,
+            format!("Failed to open file '{}': {}", image_file_path, e),
+        )
+    })?;
+
+    let mut data = Vec::new();
+    file.read_to_end(&mut data).map_err(|e| {
+        Error::new(
+            ErrorKind::InvalidData,
+            format!("Failed to read file '{}': {}", image_file_path, e),
+        )
+    })?;
+    file.read_to_end(&mut data)?;
+    FlashImage::verify_flash_image(&data)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::PROJECT_ROOT;
+    use std::fs::{self, File};
+    use std::io::Write;
+
+    /// Helper function to create a temporary file with specific content
+    fn create_temp_file(content: &[u8], file_name: &str) -> io::Result<String> {
```

Comment:
Instead of managing these files yourself, I'd recommend using `tempfile::NamedTempFile` -- it should simplify things and ensure that the files are cleaned up.

Reply:
thanks. i'll try this one.